File: JAVA/Node.java
import java.util.ArrayList;
import java.util.List;

/**
 * @author 1M0NK3Y
 * @version 1.0.0
 *
 */

public class Node {
    
    private static int nodes_created=0;

    private int noden;
    private int tag;
    private List fromdgeW = new ArrayList<Integer>();
    private List toedgeW = new ArrayList<Integer>();
    private List from = new ArrayList<Node>();
    private int noffromnode = 0;
    private List to = new ArrayList<Node>();
    private int noftonode = 0;
    private String nodename;


    public Node(String nn){

        this.noden = Node.nodes_created;
        Node.nodes_created++;
        this.nodename = nn;
        this.tag = 0;

    }

    public Node(Node f, Node t, String nn){

        this.noden = Node.nodes_created;
        Node.nodes_created++;
        this.from.add(f);
        this.noffromnode++;
        this.to.add(t);
        this.noftonode++;
        this.nodename = nn;
        this.fromdgeW.add(f.getToW(this));
        this.toedgeW.add(t.getFromW(this));
        this.tag = f.getToW(this) + f.getTag();


    }


    public int getToW(Node n){

        Node n2;

        for(int i=0; i < this.noftonode; i++){

            n2 = (Node) this.to.get(i);

            if (n.equals(n2)) {

                int g = (int) this.toedgeW.get(i);

                return g;
                
            }

        }

        return -404;

    }

    public int getFromW(Node n){

        Node n2;

        for(int i=0; i < this.noffromnode; i++){

            n2 = (Node) this.from.get(i);

            if (n.equals(n2)) {

                int g = (int) this.fromdgeW.get(i);

                return g;
                
            }

        }

        return -404;

    }

    public int getTag(){

        return this.tag;

    }

    public int getNodeN(){

        return this.noden;

    }

    public String getNodeName(){

        return this.nodename;

    }

    public Node getFromNode(int i){

        Node n = (Node) this.from.get(i);

        return n;

    }

    public Node getToNode(int i){

        Node n = (Node) this.to.get(i);

        return n;

    }


    public Node addFromNode(Node n){

        this.from.add(n);
        this.noffromnode++;
        this.fromdgeW.add(n.getToW(this));

        this.calcTag();

        for(int i=0; i < this.noftonode; i++){

            Node n2 = (Node) this.to.get(i);
            n2.calcTag();

        }

        return this;

    }

    public Node addToNode(Node n, int w){

        this.to.add(n);
        this.noftonode++;
        this.toedgeW.add(w);

        n.addFromNode(this);

        return this;

    }

    // TODO: 
    /* 
    public int setFromnNodeW(int w){

        this.fromdgeW = w;
        this.from.setToNodeW(w);
        this.calcTag();
        return this.fromdgeW;

    }

    public int setToNodeW(int w){

        this.toedgeW = w;
        return this.toedgeW;

    }

    */

    public int calcTag(){

        int mintag = 1000000, minW = 1000000, nw;
        Node n;

        for(int i=0; i < this.noffromnode; i++){

            n = (Node) this.from.get(i);
            nw = (int) this.fromdgeW.get(i);

            if((n.getTag() + nw) < (mintag + minW)){

                mintag = n.getTag();
                minW = nw;

            }

        }



        this.tag = mintag + minW;

        
        return this.tag;

    }

    
    @Override
    public boolean equals(Object obj) {
        
        if (obj == null || !(obj instanceof Node)) {

            return false;
            
        }else{

            Node n = (Node) obj;

            if (this.nodename.equals(n.getNodeName()) && this.noden == n.getNodeN() && this.tag == n.getTag()) {

                return true;
                
            }else{

                return false;

            }

        }
    }

    @Override
    public String toString() {

        return "|" + this.nodename + ":" + this.tag + "|";

    }

}

File: PYTHON/Node.py
class Node():


    def __init__(self, name):
        
        self.nodename = name;
        self.fromnodes = []
        self.fromW = []
        self.noffromnode = 0
        self.to = []
        self.toW = []
        self.noftonode = 0
        self.tag = 0;


    

    def findW(self, n):

        i = 0

        r = 1000000

        for x in self.to:

            #if x.findW(n, False) + self.toW[i] < resultT and t == True:
                
                #resultT = x.findW(n, False) + self.toW[i]


            if x.getNodeName() == n.getNodeName() and x.getFromW(self) >= 0 and x.getFromW(self) < r:

                r = self.toW[i]

            i += 1

        i = 0
        for x in self.fromnodes:

            #if x.findW(n, False) + self.fromW[i] < resultF and t == True:

                #resultF = x.findW(n, False) + self.fromW[i]


            if x.getNodeName == n.getNodeName() and x.getToW(self) >= 0 and x.getToW(n) < r:

                r = self.fromW[i]

            i += 1



        i = 0

        for x in self.fromnodes:
        

            if x.isConnected(n) == -1 and x.getFromW(n) > -1:


                if x.getFromW(n) + self.fromW[i] < r and r >= 0 and x.getFromW(n) + self.fromW[i] >= 0:

                    r = x.getFromW(n) +self.fromW[i]
        


            elif x.isConnected(n) == 1 and x.getToW(n) > -1:

                if x.getToW(n) + self.fromW[i] < r and r >= 0 and x.getToW(n) + self.fromW[i] >= 0:

                    r = x.getToW(n) + self.fromW[i]
                

            i += 1


        i = 0

        for x in self.to:

            if x.isConnected(n) == -1 and x.getFromW(n) > -1:

                if x.getFromW(n) + self.toW[i] < r and r >= 0 and x.getFromW(n) + self.toW[i] >= 0:

                    r = x.getFromW(n) + self.toW[i]

            elif x.isConnected(n) == 1 and x.getToW(n) > -1:

                if x.getToW(n) + self.toW[i] < r and r >= 0 and x.getToW(n) + self.toW[i] >= 0:

                    r = x.getFromW(n) + self.toW[i]

            i += 1
        
        if r != 1000000:

            return r

        else:

            return -1


    def getFromW(self, n):

        i = 0
        
        for x in self.fromnodes:

            if x.equals(n):

                return self.fromW[i]
            
            i += 1

        return -1

    def getToW(self, n):

        i = 0

        for x in self.to:

            if x.equals(n):

                return self.toW[i]

            i += 1

        return -1

    def isConnected(self, n):

        for x in self.to:

            if x.equals(self):

                return 1

        for x in self.fromnodes:

            if x.equals(self):

                return -1

        return 0


    def getTag(self):

        return self.tag


    def getNodeName(self):

        return self.nodename

    def getFromNode(self, i):

        return self.fromnodes[i]

    def getToNode(self, i):

        return self.to[i]

    def setWFromNode(self, i, w):

        self.fromW[i] = w

        return self.fromnodes

    def setWToNode(self, i, w):

        self.toW[i] = w

        return self.toW[i]


    def addFromNode(self, n, w):

        self.fromnodes.append(n)
        self.fromW.append(w)
        self.noffromnode += 1

        self.calcTag()
    
    def addToNode(self, n, w):

        self.to.append(n)
        self.toW.append(w);
        self.noftonode += 1

        n.addFromNode(self, w)


    def calcTag(self):

        mintag = 1000000
        minw = mintag
        i = 0

        while i < self.noffromnode:

            if self.fromnodes[i].getTag() >= 0 and self.fromW[i] >= 0 and ((self.fromnodes[i].getTag() + self.fromW[i]) < (mintag + minw)) and self.fromnodes[i].getNodeName() != self.nodename:

                mintag = self.fromnodes[i].getTag()
                minw = self.fromW[i]

            i += 1


        self.tag = mintag + minw
        
        return self.tag


    def equals(self, n):

        if self.tag == n.getTag() and self.nodename == n.getNodeName():

            return True

        else:

            return False


    def toString(self):

        s = f"|{self.nodename}:{self.tag}|"

        return s


